### ISSUE TITLE
Add financial year start date option for budget

### ISSUE THREAD

**Issue by MarcusGHub** [moneymanagerex/moneymanagerex#1292]
A financial month would be beneficial.

I get my salary on the 23rd of each month. So, my financial month would begin on the 24th of each month and end on the 23rd of the following month. However, with the current budget feature in Money Manager Ex, an entire month is used.

I work salary to salary, so having a regular monthly budget is a no no for me.

With this feature, my budget would like this:
January budget - Dec 24-Jan 23
February budget  - Jan 24 - Feb 23
March budget  - Feb 24 - Mar 23
.
.
.
December budget  - Nov 24 - Dec 23

My financial year would begin at December 24th.

Can't we get an option to select our financial month, like how we select our financial year? That's the only feature this amazing app is missing.

Thanks

**Comment by slodki** [comment]
The Jan 15 to Feb 14 period should be named January or February budget? This is not clear for me - you must compare your **January** salary to your spending in this period...

**Comment by MarcusGHub** [comment]
Sorry for not being clear enough... let me try again.

**CURRENT FEATURE**
So as it stands, budgeting in MMEX can be set to yearly or monthly. I prefer monthly budgets.
Monthly budget runs from the first day of the month to the last.
So, a budget for January would be called
2017 - 01
The dates would be January 1 - January 31

A budget for February would be called
2017 - 02
The dates would be February 1 - February 28

A budget for March would be called
2017 - 03
The dates would be March 1 - March 31

**FEATURE REQUEST**
I cannot work with a budget with the dates above. I would like a feature to change the date range. So instead of the date starting at the beginning of the month, I can set a start date at the 24th of the month and end date at the 23rd of the following month. This would be really beneficial because I budget from the day after payday to the following payday.

A budget for January would be called
2017 - 01
The dates would be December 24 - January 23

A budget for February would be called
2017 - 02
The dates would be January 24 - February 23

A budget for March would be called
2017 - 03
The dates would be February 24 - March 23

**Comment by slodki** [comment]
My qestion was about name only - when budget civers one calendar month then name is obvious.  But why the half January/half February budget is named January not February?

It's quite common to start financial year after school holidays etc. and your idea is clear to allow change start budget year (and month) at any date selected by user.

**Comment by MarcusGHub** [comment]
See I was off again :)
I understand now. 

So I get paid on the 23rd of each month. So I would like to budget from the 24th of that month to the 23rd of the next month.
March budget would be from Feb 24-Mar 23.
The spending would be mostly for the month of March.

April budget would be from Mar 24-Apr 23.
This spending would be for the month of April.

I have been budgeting like this since 2015 with Excel, but MMEX is leagues ahead.

**Comment by MarcusGHub** [comment]
Please tell me this is being considered *pwease*

**Comment by stef145g** [comment]
We have a financial year setting for accounts. This would have to be a specific setting for budgets such as "Budget Offset Date" in order not to conflict with Financial Year settings. I will investigate this possibility.

**Comment by MarcusGHub** [comment]
Hi
How is this coming along?

**Comment by MarcusGHub** [comment]
@stef145g any update?

**Comment by stef145g** [comment]
Family issues and other commitments have delayed my investigation and implementation.

I have now been able to add a new Option on the Views page called: **Budget Offset  (days)**
This option changes the start date of the 1st of the month or year for a budget by the budget offset in days. This means the budget start can be set forward or back up to 30 days.

@MarcusGHub Please try out the new feature with the build for Windows found here:
https://ci.appveyor.com/project/stef145g/moneymanagerex/build/1185/job/ms61lgfoh8rre8qq/artifacts

When all is well, I will add it to moneymanagerex/master

**Comment by MarcusGHub** [comment]
@stef145g THANK YOU a lot!!! I'm sooo grateful!
Sorry for shouting - lol

So I enabled the option as you instructed.
I initially set it to offset by 23 days. It worked as it should. However, I realized that if I select the month of September, the budget would cover from September 24-October ##. The budget would be titled 2017-09, and I would consider it to be for September. However, I would prefer my budget for September be from August 24-September 23.

So I changed the offset days to -9, instead of 23. This gave me what I wanted.

My question is; how accurate is this?
How does it account for months with varying number of days, like 28, 30 and 31 days?

**Comment by slodki** [comment]
Why the 30 days limit? Can I offset year budget to April 5th?

Negative numbers can be solution to select naming shifted periods:
- if offset is positive number OR ZERO then period is named on first day of period (budget month name, budget year etc.)
- if offset is negative number then period is named on last day of period

Examples:
- offset=0
  - 2017-05-01 - 2017-05-31 is May budget period
  - 2017-01-01 - 2017-12-31 is 2017 budget period
- offset=5
  - 2017-05-06 - 2017-06-05 is May budget period
  - 2017-01-06 - 2018-01-05 is 2017 budget period
- offset=-26
  - 2017-05-06 - 2017-06-05 is June budget period
  - 2017-01-06 - 2018-01-05 is 2018 budget period

**Comment by stef145g** [comment]
The Financial Year setting alters the date of a financial year which also affects a budget so it does not have to start on January 1st of a year. This budget date shift is only supposed to be within a short period to align paydays for budgeting purposes. Example: Financial year starts on 1st July but the start of the pay week starts on the 26th June, but you get paid on the 3rd July which is the new financial year.  The 30 day limit is to restrict the date so it is within a month. Having a shift beyond this is not appropriate.  Anything more that this can cause confusion.

As for accuracy, the dates are calculated on the initial year or month settings. The shift affects both the start and end date of a period so the accuracy should not be affected.

**Comment by MarcusGHub** [comment]
@stef145g was this feature added to main app as yet, so I can move over and still get updates?

**Comment by stef145g** [comment]
@MarcusGHub This feature has been added to main, but has not been set in any release as yet.

To be able to use this feature you can download the latest build for Windows from [Appveyor](https://ci.appveyor.com/project/moneymanagerex/moneymanagerex). Instructions on how to navigate Appveyor can be obtained [here](https://github.com/moneymanagerex/moneymanagerex/wiki/How-to-download-the-mmex-build)

**Comment by MarcusGHub** [comment]
@stef145g was this adjustment done on the basis that a month can have 31 days, and the offset affect that number and not the fact that each month has their own set amount of days?

This offset has been working well up until the new budget date of February 23.  Budget offset is set to -9. Budget for 2018-03 actually started on the 21st instead of the 23rd/24th. I know any changes that I make to the offset days number that is working up until now will affect past budgets.

**Comment by Pow377** [comment]
This change would need to either have a start date set throughout budgets, or allow an offset per budget.
I am paid on the 18th each month, so need to budget 18th - 17th.

Each month the offset will change (as different number of days in the month).

**Comment by stef145g** [comment]
The adjustment was a direct add or subtract x days from the first of the month. If you need the 18th of each month then adding 18 to each month should achieve this. If it is a particular day in a positive direction eg. Tuesday in the 3rd week, then a different adjustment needs to be worked out, same as in a negative direction.

**Comment by Pow377** [comment]
Ah apologies I must have misunderstood how it works, I looked at the  "Start Date of:" label but was always empty.
Thanks for confirming.

**Comment by MarcusGHub** [comment]
@stef145g.... I like the idea of 18th of each month. Setting a day for each month would work for me.
The only month that would differ would be January, due to the fact that I get paid earlier that usual because of the holidays.

**Comment by dv0zn3r** [comment]
Is there any way to offset the date for monthly budget in OS X version of MMEX?
